File: mvp/sdk/typescript/src/api/knowledge.ts
import { BaseAPI } from "./base";
import {
  Document,
  DocumentChunk,
  KnowledgeSearchRequest,
  KnowledgeSearchResponse,
  AskRequest,
  AskResponse,
  SearchStatus,
} from "../types";

/**
 * Knowledge Base API — document upload, chunking, hybrid search
 * (pgvector + TF-IDF), and RAG question answering.
 */
export class KnowledgeAPI extends BaseAPI {
  // -- Documents -----------------------------------------------------------

  /** Upload a document (TXT, MD, CSV; max 10 MB). */
  async upload(file: Blob | File): Promise<Document> {
    const form = new FormData();
    form.append("file", file);
    return this._request<Document>("/api/knowledge/upload", {
      method: "POST",
      body: form,
    });
  }

  /** List user documents. */
  async listDocuments(): Promise<Document[]> {
    return this._get<Document[]>("/api/knowledge/documents");
  }

  /** List chunks for a document. */
  async listChunks(documentId: string): Promise<DocumentChunk[]> {
    return this._get<DocumentChunk[]>(`/api/knowledge/documents/${documentId}/chunks`);
  }

  /** Delete a document and all its chunks. */
  async deleteDocument(documentId: string): Promise<void> {
    return this._delete(`/api/knowledge/documents/${documentId}`);
  }

  // -- Search --------------------------------------------------------------

  /** Hybrid search (auto-detects best mode: TF-IDF, vector, or hybrid). */
  async search(data: KnowledgeSearchRequest): Promise<KnowledgeSearchResponse> {
    return this._post<KnowledgeSearchResponse>("/api/knowledge/search", data);
  }

  /** Vector-only search using pgvector. */
  async vectorSearch(data: KnowledgeSearchRequest): Promise<KnowledgeSearchResponse> {
    return this._post<KnowledgeSearchResponse>("/api/knowledge/search/vector", data);
  }

  /** Get available search modes and their status. */
  async searchStatus(): Promise<SearchStatus> {
    return this._get<SearchStatus>("/api/knowledge/search/status");
  }

  // -- RAG -----------------------------------------------------------------

  /** Ask a question — searches knowledge base and generates an AI answer. */
  async ask(data: AskRequest): Promise<AskResponse> {
    return this._post<AskResponse>("/api/knowledge/ask", data);
  }

  // -- Maintenance ---------------------------------------------------------

  /** Reindex all chunks with fresh embeddings. */
  async reindexEmbeddings(): Promise<{ status: string; reindexed: number }> {
    return this._post<{ status: string; reindexed: number }>("/api/knowledge/reindex-embeddings");
  }
}

File: mvp/sdk/python/saas_ia/api/knowledge.py
"""Knowledge Base API — document upload, chunking, hybrid search, RAG."""

from __future__ import annotations

from typing import Any, BinaryIO

from saas_ia.api.base import BaseAPI


class KnowledgeAPI(BaseAPI):
    """Wraps ``/api/knowledge`` endpoints."""

    # -- Documents ----------------------------------------------------------

    async def upload(
        self,
        file: BinaryIO,
        filename: str = "upload.txt",
    ) -> dict[str, Any]:
        """Upload a document (TXT, MD, CSV; max 10 MB)."""
        return await self._post(
            "/api/knowledge/upload",
            files={"file": (filename, file)},
        )

    async def list_documents(self) -> list[dict[str, Any]]:
        """List user documents."""
        return await self._get("/api/knowledge/documents")

    async def list_chunks(self, document_id: str) -> list[dict[str, Any]]:
        """List chunks for a document."""
        return await self._get(f"/api/knowledge/documents/{document_id}/chunks")

    async def delete_document(self, document_id: str) -> None:
        """Delete a document and all its chunks."""
        await self._delete(f"/api/knowledge/documents/{document_id}")

    # -- Search -------------------------------------------------------------

    async def search(
        self,
        query: str,
        *,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Hybrid search (auto-detects best mode)."""
        body: dict[str, Any] = {"query": query}
        if limit is not None:
            body["limit"] = limit
        return await self._post("/api/knowledge/search", json=body)

    async def vector_search(
        self,
        query: str,
        *,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Vector-only search using pgvector."""
        body: dict[str, Any] = {"query": query}
        if limit is not None:
            body["limit"] = limit
        return await self._post("/api/knowledge/search/vector", json=body)

    async def search_status(self) -> dict[str, Any]:
        """Get available search modes and their status."""
        return await self._get("/api/knowledge/search/status")

    # -- RAG ----------------------------------------------------------------

    async def ask(
        self,
        question: str,
        *,
        limit: int | None = None,
    ) -> dict[str, Any]:
        """Ask a question -- searches knowledge base and generates an AI answer."""
        body: dict[str, Any] = {"question": question}
        if limit is not None:
            body["limit"] = limit
        return await self._post("/api/knowledge/ask", json=body)

    # -- Maintenance --------------------------------------------------------

    async def reindex_embeddings(self) -> dict[str, Any]:
        """Reindex all chunks with fresh embeddings."""
        return await self._post("/api/knowledge/reindex-embeddings")
